Cwd proposal delegate

baoskee opened the pull request:
Need design review before I start writing tests! : ) 
Low prio, no hurries. Thought would be fun first contract to write for DaoDao.

baoskee committed:
init cwd-proposal-delegate

baoskee committed:
Msg + Error + State design

baoskee committed:
Skeleton code setup

baoskee committed:
execute and state redesign for security

baoskee committed:
execute_delegate and execute_remove_delegation

baoskee committed:
execute_execute body

baoskee committed:
Reply implementation for submessage

baoskee committed:
Query functions

baoskee committed:
Added smart contract risks section

0xekez commented on contracts/proposal/cwd-proposal-delegate/Cargo.toml:
```diff
@@ -0,0 +1,55 @@
+# use library feature to disable all instantiate/execute/query exports
+library = []
+
+[package.metadata.scripts]
+optimize = """docker run --rm -v "$(pwd)":/code \
+  --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target \
+  --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry \
+  cosmwasm/rust-optimizer:0.12.6
+"""
```
because this needs to be built as part of the workspace (you have a cwd-core dep) this won't work / we don't need it
```suggestion
```

0xekez commented on contracts/proposal/cwd-proposal-delegate/Cargo.toml:
```diff
@@ -0,0 +1,55 @@
+[dependencies]
+cosmwasm-schema = "1.1.3"
+cosmwasm-std = "1.1.3"
+cosmwasm-storage = "1.1.3"
```
almost no contracts depend on this, yet it is part of the default template lol
```suggestion
```

0xekez commented on contracts/proposal/cwd-proposal-delegate/Cargo.toml:
```diff
@@ -0,0 +1,55 @@
+  cosmwasm/rust-optimizer:0.12.6
+"""
+
+[dependencies]
```
all of these can be `cosmwasm-x = { workspace = true } `

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/contract.rs:
```diff
@@ -0,0 +1,238 @@
+        REPLY_ID_EXECUTE_PROPOSAL_HOOK => {
+            let id = EXECUTE_CTX.load(deps.storage)?;
+            let Delegation {
+                policy_preserve_on_failure,
```
this is sick

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/contract.rs:
```diff
@@ -0,0 +1,238 @@
+                ..
+            } = DELEGATIONS.load(deps.storage, id)?;
+
+            if let SubMsgResult::Err(_) = msg.result {
+                // && with `let` expression above gives unstable Rust error
+                if policy_preserve_on_failure {
```
```suggestion
            if msg.result.is_err() && policy_preserve_on_failure {
```

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/contract.rs:
```diff
@@ -0,0 +1,238 @@
+                        .add_attribute("execute_failed_but_preserved", id.to_string()));
+                }
+            }
+            // Delete delegation in both success and error case
+            DELEGATIONS.remove(deps.storage, id);
+            Ok(Response::default())
```
nit: nice to have this in an `else` block because it makes it clear from reading that we'll never "fall out" of the if statement above and into this.

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/contract.rs:
```diff
@@ -0,0 +1,238 @@
+    Ok(DelegationCountResponse { count })
+}
+
+fn query_delegation(deps: Deps, delegation_id: u64) -> StdResult<DelegationResponse> {
+    let delegation = DELEGATIONS.load(deps.storage, delegation_id)?;
+    Ok(delegation)
+}
```
probably want to make this a paginated function so that FE can list all of the delegations pending. `packages/cw-paginate` should make that simple

baoskee replied:
cool!

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/contract.rs:
```diff
@@ -0,0 +1,238 @@
+
+// MARK: Query helpers
+
+fn query_delegation_count(deps: Deps) -> StdResult<DelegationCountResponse> {
+    let count = DELEGATION_COUNT.load(deps.storage)?;
+    Ok(DelegationCountResponse { count })
+}
```
idea: why expose this at all? all we use it for is a nonce for delegations. let's make it a private state variable and only expose a function for incrementing it. gets us pretty close to correct-by-constructuon!

ex:

```rust
pub(crate) fn advance_current_id(store: &mut dyn Storage) -> StdResult<u64> {
    let CURRENT_ID: Item<u64> = Item::new("current_id");
    let id: u64 = CURRENT_ID.may_load(store)?.unwrap_or_default() + 1;
    CURRENT_ID.save(store, &id)?;
    Ok(id)
}
```

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/contract.rs:
```diff
@@ -0,0 +1,238 @@
+    if info.sender != admin {
+        return Err(ContractError::Unauthorized {});
+    }
+    assert_not_expired(&delegation.expiration, &env.block)?;
```
nice check

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/contract.rs:
```diff
@@ -0,0 +1,238 @@
+    }
+    // If delegation is irrevocable, return Error
+    let Delegation {
+        policy_irrevocable, ..
```
what's the reason for having this option? i feel things like this can give fake guarantees if the contract is instantiated with a DAO as it's admin. should we be instantiating this without an admin perhap?

baoskee replied:
good point. The DAO could always remove the proposal module and all delegations would be revocable. However, that would eliminate all their existing delegations as well. There's 2 options to go with this:
1. Rename the policy to `policy_proposal_irrevocable` and add a warning
2. Remove the policy altogether

0xekez commented on contracts/proposal/cwd-proposal-delegate/src/error.rs:
```diff
@@ -0,0 +1,21 @@
+
+    #[error("Unauthorized")]
+    Unauthorized {},
+    // Add any other custom errors you like here.
+    // Look at https://docs.rs/thiserror/1.0.21/thiserror/ for details.
```
```suggestion
```

0xekez reviewed (commented):
🫡 cool! super clean. let's land this :)

baoskee commented:
Will integrate changes and start on the tests! A bit slow on this, been grokking some Cosmos stuff. Need to learn Go, how dreadful

baoskee committed:
Update contracts/proposal/cwd-proposal-delegate/Cargo.toml

Co-authored-by: ekez <30676292+0xekez@users.noreply.github.com>

baoskee committed:
Update contracts/proposal/cwd-proposal-delegate/Cargo.toml

Co-authored-by: ekez <30676292+0xekez@users.noreply.github.com>

baoskee committed:
Merge https://github.com/DA0-DA0/dao-contracts into cwd-proposal-delegate

baoskee committed:
new cargo.toml

baoskee committed:
Removed exposure of delegation count

baoskee committed:
merged

baoskee committed:
test shell

baoskee committed:
Added testing strategy

baoskee committed:
test unauthorized delegation

baoskee committed:
execute_authorization tests

baoskee committed:
Expiration test

baoskee committed:
Revocable policy and delegation not found error path

baoskee committed:
First try at testing submessage

baoskee committed:
Added pagination query for delegations

baoskee committed:
Reply hook testing

baoskee committed:
refactor to policy_module_irrevocable

baoskee commented:
Ok, last commits were:
- fixed Cargo.toml
- made changes for PR
- added pagination query (can remove the single delegation query if that's cleaner)
- made delegation count private
- renamed policy_revocable to `policy_module_irrevocable`. I think that having revocability at the module level is convenient even though it can expose user error (because DAO can remove the module to "revoke" the delegation). For example, the Constitution can technically disband the Supreme Court to prevent a particular judge-made decision to not be implemented (say McCulloch v. Maryland), but that would require high costs plus the dismantling of every other Court decision
- added a bunch of unit tests covering the risks I laid out in README.md 

Lmk.

JakeHartnell commented:
Closing as stale.
